### PR TITLE
extensions: add basic support

### PR DESCRIPTION
**Pull request by tigarmo**
Much like Snapcraft's, Extensions can add top-level elements, new parts,
or modify existing parts in Rockcraft projects. This commit implements
the basic support of validating extensions and applying them when the
project is loaded.

--------

This PR's history is messy because I created this branch off of the existing `extensions-flask`, merged main and then removed the code relating to the command line and to flask (those will be addressed separately). @sergiusens db58974f0560a3 is the one with my changes and tests.

### CODE REVIEW

**Review by cjdcordeiro — COMMENTED** [review]
does this deserve a reference in the docs and maybe a spread test?

**Comment by tigarmo** [comment]
It does deserve being documented and spread-tested, but I think we should wait until we have at least one extension working. With this single PR not even the command line options exist, so we would be documenting a very "abstract" idea (and there would be nothing to test unless we add some kind of "fake" extension solely for testing)

File: tests/unit/extensions/test_extensions.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        services:
+            my-service:
+                command: foo
+                override: merge
+    """
+    )
+    rockcraft_yaml = tmp_path / "rockcraft.yaml"
+    rockcraft_yaml.write_text(yaml_data)
+
+    project = load_project(rockcraft_yaml)
```
[the remark this answers is not in the document]
Comment (tigarmo) [reply tests/unit/extensions/test_extensions.py:216]: can you explain what you mean here? I don't get it

[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/unit/extensions/test_extensions.py:216]: oh, just a side comment for improving load_project to not require a Path().write_text() call

File: rockcraft/extensions/_utils.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+            return merged
+
+        if isinstance(existing_property, dict) and isinstance(extension_property, dict):
+            for key, value in extension_property.items():
+                existing_property[key] = _apply_extension_property(
+                    existing_property.get(key), value
+                )
+            return existing_property
+        return existing_property
```
Comment (cmatsuoka) [inline rockcraft/extensions/_utils.py:100]: We could remove the inner return here.

**Review by cmatsuoka — APPROVED** [review]
Looks good, added a couple of nitpicks.

**Pull request merged**